-Wuseless-cast warning in ASSERT with GCC 13

I got the following `-Wuesless-cast` warning with GCC 13:

```
In file included from cpp-sort/include/cpp-sort/detail/splaysort.h:21,
                 from cpp-sort/include/cpp-sort/sorters/splay_sorter.h:21,
                 from cpp-sort/include/cpp-sort/sorters.h:33,
                 from cpp-sort\examples\readme_example.cpp:11:
cpp-sort/include/cpp-sort/detail/splaysort.h: In member function 'void cppsort::detail::splay_tree_base::splay(cppsort::detail::binary_tree_node_base*)':
libassert-src/include/assert.hpp:1253:45: warning: useless cast to type 'struct libassert::detail::expression_decomposer<cppsort::detail::binary_tree_node_base*&, cppsort::detail::binary_tree_node_base*, libassert::detail::ops::neq>' [-Wuseless-cast]
 1253 |                          libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
      |                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libassert-src/include/assert.hpp:1171:53: note: in definition of macro 'LIBASSERT_STMTEXPR'
 1171 |  #define LIBASSERT_STMTEXPR(B, R) (__extension__ ({ B R }))
      |                                                     ^
libassert-src/include/assert.hpp:1293:28: note: in expansion of macro 'ASSERT_INVOKE'
 1293 |  #define ASSERT(expr, ...) ASSERT_INVOKE(expr, true, true, "ASSERT", assertion, , __VA_ARGS__)
      |                            ^~~~~~~~~~~~~
cpp-sort/include/cpp-sort/detail/splaysort.h:80:17: note: in expansion of macro 'ASSERT'
   80 |                 ASSERT(parent != &sentinel_node_);
```

It occurred while comparing variables of pointer type. I am still trying to produce a minimum viable example to narrow it down.

All I need to reproduce the warning with the code above is to call CMake with `-DASSERT_DECOMPOSE_BINARY_LOGICAL=ON` and to enable `-Wuseless-cast`.

Which gives the following warning:

```
In file included from test.cpp:1:
test.cpp: In function 'int main()':
assert.hpp:1253:45: warning: useless cast to type 'struct libassert::detail::expression_decomposer<int&, int&, libassert::detail::ops::neq>' [-Wuseless-cast]
 1253 |                          libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
      |                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
assert.hpp:1171:53: note: in definition of macro 'LIBASSERT_STMTEXPR'
 1171 |  #define LIBASSERT_STMTEXPR(B, R) (__extension__ ({ B R }))
      |                                                     ^
assert.hpp:1293:28: note: in expansion of macro 'ASSERT_INVOKE'
 1293 |  #define ASSERT(expr, ...) ASSERT_INVOKE(expr, true, true, "ASSERT", assertion, , __VA_ARGS__)
      |                            ^~~~~~~~~~~~~
test.cpp:6:5: note: in expansion of macro 'ASSERT'
    6 |     ASSERT(a != b);
      |     ^~~~~~
```

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1170,10 +1170,14 @@ using libassert::ASSERTION;
  // with decltype(auto) in an expression like decltype(auto) x = __extension__ ({...}).y;
  #define LIBASSERT_STMTEXPR(B, R) (__extension__ ({ B R }))
  #define LIBASSERT_WARNING_PRAGMA_PUSH _Pragma("GCC diagnostic push")
- #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
-    _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
-    _Pragma("GCC diagnostic ignored \"-Wunknown-warning-option\"") /* TODO: Awful */ \
-    _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
+ #if LIBASSERT_IS_GCC
+  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
+     _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
+     _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
+ #else
+  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
+     _Pragma("GCC diagnostic ignored \"-Wparentheses\"")
+ #endif
  #define LIBASSERT_WARNING_PRAGMA_POP _Pragma("GCC diagnostic pop")
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) static_cast<bool>(x)
 #else

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1169,11 +1169,17 @@ using libassert::ASSERTION;
  // Extra set of parentheses here because clang treats __extension__ as a low-precedence unary operator which interferes
  // with decltype(auto) in an expression like decltype(auto) x = __extension__ ({...}).y;
  #define LIBASSERT_STMTEXPR(B, R) (__extension__ ({ B R }))
- #define LIBASSERT_WARNING_PRAGMA _Pragma("GCC diagnostic ignored \"-Wparentheses\"")
+ #define LIBASSERT_WARNING_PRAGMA_PUSH _Pragma("GCC diagnostic push")
+ #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
+    _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
+    _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
+ #define LIBASSERT_WARNING_PRAGMA_POP _Pragma("GCC diagnostic pop")
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) static_cast<bool>(x)
 #else
  #define LIBASSERT_STMTEXPR(B, R) [&](const char* libassert_msvc_pfunc) { B return R }(LIBASSERT_PFUNC)
- #define LIBASSERT_WARNING_PRAGMA
+ #define LIBASSERT_WARNING_PRAGMA_PUSH
+ #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA
+ #define LIBASSERT_WARNING_PRAGMA_POP
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) libassert::detail::static_cast_to_bool(x)
  namespace libassert::detail {
      template<typename T> constexpr bool static_cast_to_bool(T&& t) {
@@ -1248,9 +1254,11 @@ using libassert::ASSERTION;
 #define ASSERT_INVOKE(expr, doreturn, check_expression, name, type, failaction, ...) \
         LIBASSERT_IGNORE_UNUSED_VALUE \
         LIBASSERT_STMTEXPR( \
-          LIBASSERT_WARNING_PRAGMA \
+          LIBASSERT_WARNING_PRAGMA_PUSH \
+          LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
           auto libassert_decomposer = \
                          libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
+          LIBASSERT_WARNING_PRAGMA_POP \
           decltype(auto) libassert_value = libassert_decomposer.get_value(); \
           constexpr bool libassert_ret_lhs = libassert_decomposer.ret_lhs(); \
           if constexpr(check_expression) { \

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1287,7 +1287,6 @@ using libassert::ASSERTION;
                 new (&libassert_decomposer) libassert::detail::expression_decomposer(std::move(libassert_r)); \
               } \
             } \
-        LIBASSERT_WARNING_PRAGMA_POP \
           }, \
           /* Note: std::launder needed in 17 in case of placement new / move shenanigans above */ \
           /* https://timsong-cpp.github.io/cppwp/n4659/basic.life#8.3 */ \
@@ -1295,7 +1294,8 @@ using libassert::ASSERTION;
           libassert::detail::get_expression_return_value <doreturn LIBASSERT_COMMA \
             libassert_ret_lhs LIBASSERT_COMMA std::is_lvalue_reference<decltype(libassert_value)>::value> \
               (libassert_value, *std::launder(&libassert_decomposer)); \
-        ) LIBASSERT_IF(doreturn)(.value,)
+        ) LIBASSERT_IF(doreturn)(.value,) \
+        LIBASSERT_WARNING_PRAGMA_POP
 
 #ifdef NDEBUG
  #define LIBASSERT_ASSUME_ACTION LIBASSERT_UNREACHABLE;

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1287,6 +1287,7 @@ using libassert::ASSERTION;
                 new (&libassert_decomposer) libassert::detail::expression_decomposer(std::move(libassert_r)); \
               } \
             } \
+        LIBASSERT_WARNING_PRAGMA_POP \
           }, \
           /* Note: std::launder needed in 17 in case of placement new / move shenanigans above */ \
           /* https://timsong-cpp.github.io/cppwp/n4659/basic.life#8.3 */ \
@@ -1294,7 +1295,6 @@ using libassert::ASSERTION;
           libassert::detail::get_expression_return_value <doreturn LIBASSERT_COMMA \
             libassert_ret_lhs LIBASSERT_COMMA std::is_lvalue_reference<decltype(libassert_value)>::value> \
               (libassert_value, *std::launder(&libassert_decomposer)); \
-        LIBASSERT_WARNING_PRAGMA_POP \
         ) LIBASSERT_IF(doreturn)(.value,)
 
 #ifdef NDEBUG

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1169,23 +1169,31 @@ using libassert::ASSERTION;
  // Extra set of parentheses here because clang treats __extension__ as a low-precedence unary operator which interferes
  // with decltype(auto) in an expression like decltype(auto) x = __extension__ ({...}).y;
  #define LIBASSERT_STMTEXPR(B, R) (__extension__ ({ B R }))
- #define LIBASSERT_WARNING_PRAGMA_PUSH _Pragma("GCC diagnostic push")
  #if LIBASSERT_IS_GCC
   #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
      _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
      _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
+  #define LIBASSERT_WARNING_PRAGMA_PUSH_GCC _Pragma("GCC diagnostic push")
+  #define LIBASSERT_WARNING_PRAGMA_POP_GCC _Pragma("GCC diagnostic pop")
+  #define LIBASSERT_WARNING_PRAGMA_PUSH_CLANG
+  #define LIBASSERT_WARNING_PRAGMA_POP_CLANG
  #else
   #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
      _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
      _Pragma("GCC diagnostic ignored \"-Woverloaded-shift-op-parentheses\"")
+  #define LIBASSERT_WARNING_PRAGMA_PUSH_GCC
+  #define LIBASSERT_WARNING_PRAGMA_POP_GCC
+  #define LIBASSERT_WARNING_PRAGMA_PUSH_CLANG _Pragma("GCC diagnostic push")
+  #define LIBASSERT_WARNING_PRAGMA_POP_CLANG _Pragma("GCC diagnostic pop")
  #endif
- #define LIBASSERT_WARNING_PRAGMA_POP _Pragma("GCC diagnostic pop")
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) static_cast<bool>(x)
 #else
  #define LIBASSERT_STMTEXPR(B, R) [&](const char* libassert_msvc_pfunc) { B return R }(LIBASSERT_PFUNC)
- #define LIBASSERT_WARNING_PRAGMA_PUSH
+ #define LIBASSERT_WARNING_PRAGMA_PUSH_CLANG
+ #define LIBASSERT_WARNING_PRAGMA_POP_CLANG
+ #define LIBASSERT_WARNING_PRAGMA_PUSH_GCC
+ #define LIBASSERT_WARNING_PRAGMA_POP_GCC
  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA
- #define LIBASSERT_WARNING_PRAGMA_POP
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) libassert::detail::static_cast_to_bool(x)
  namespace libassert::detail {
      template<typename T> constexpr bool static_cast_to_bool(T&& t) {
@@ -1259,12 +1267,14 @@ using libassert::ASSERTION;
 #endif
 #define ASSERT_INVOKE(expr, doreturn, check_expression, name, type, failaction, ...) \
         /* must push/pop out here due to nasty clang bug https://github.com/llvm/llvm-project/issues/63897 */ \
-        LIBASSERT_WARNING_PRAGMA_PUSH \
-        LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
+        LIBASSERT_WARNING_PRAGMA_PUSH_CLANG \
+        LIBASSERT_WARNING_PRAGMA_PUSH_GCC \
         LIBASSERT_IGNORE_UNUSED_VALUE \
+        LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
         LIBASSERT_STMTEXPR( \
           auto libassert_decomposer = \
                          libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
+          LIBASSERT_WARNING_PRAGMA_POP_GCC \
           decltype(auto) libassert_value = libassert_decomposer.get_value(); \
           constexpr bool libassert_ret_lhs = libassert_decomposer.ret_lhs(); \
           if constexpr(check_expression) { \
@@ -1295,7 +1305,7 @@ using libassert::ASSERTION;
             libassert_ret_lhs LIBASSERT_COMMA std::is_lvalue_reference<decltype(libassert_value)>::value> \
               (libassert_value, *std::launder(&libassert_decomposer)); \
         ) LIBASSERT_IF(doreturn)(.value,) \
-        LIBASSERT_WARNING_PRAGMA_POP
+        LIBASSERT_WARNING_PRAGMA_POP_CLANG
 
 #ifdef NDEBUG
  #define LIBASSERT_ASSUME_ACTION LIBASSERT_UNREACHABLE;

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1176,7 +1176,8 @@ using libassert::ASSERTION;
      _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
  #else
   #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
-     _Pragma("GCC diagnostic ignored \"-Wparentheses\"")
+     _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
+     _Pragma("GCC diagnostic ignored \"-Woverloaded-shift-op-parentheses\"")
  #endif
  #define LIBASSERT_WARNING_PRAGMA_POP _Pragma("GCC diagnostic pop")
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) static_cast<bool>(x)

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1172,6 +1172,7 @@ using libassert::ASSERTION;
  #define LIBASSERT_WARNING_PRAGMA_PUSH _Pragma("GCC diagnostic push")
  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
     _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
+    _Pragma("GCC diagnostic ignored \"-Wunknown-warning-option\"") /* TODO: Awful */ \
     _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
  #define LIBASSERT_WARNING_PRAGMA_POP _Pragma("GCC diagnostic pop")
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) static_cast<bool>(x)

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1258,13 +1258,13 @@ using libassert::ASSERTION;
  #endif
 #endif
 #define ASSERT_INVOKE(expr, doreturn, check_expression, name, type, failaction, ...) \
+        /* must push/pop out here due to nasty clang bug https://github.com/llvm/llvm-project/issues/63897 */ \
+        LIBASSERT_WARNING_PRAGMA_PUSH \
+        LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
         LIBASSERT_IGNORE_UNUSED_VALUE \
         LIBASSERT_STMTEXPR( \
-          LIBASSERT_WARNING_PRAGMA_PUSH \
-          LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
           auto libassert_decomposer = \
                          libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
-          LIBASSERT_WARNING_PRAGMA_POP \
           decltype(auto) libassert_value = libassert_decomposer.get_value(); \
           constexpr bool libassert_ret_lhs = libassert_decomposer.ret_lhs(); \
           if constexpr(check_expression) { \
@@ -1294,7 +1294,8 @@ using libassert::ASSERTION;
           libassert::detail::get_expression_return_value <doreturn LIBASSERT_COMMA \
             libassert_ret_lhs LIBASSERT_COMMA std::is_lvalue_reference<decltype(libassert_value)>::value> \
               (libassert_value, *std::launder(&libassert_decomposer)); \
-        ) LIBASSERT_IF(doreturn)(.value,)
+        ) LIBASSERT_IF(doreturn)(.value,) \
+        LIBASSERT_WARNING_PRAGMA_POP
 
 #ifdef NDEBUG
  #define LIBASSERT_ASSUME_ACTION LIBASSERT_UNREACHABLE;

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1294,8 +1294,8 @@ using libassert::ASSERTION;
           libassert::detail::get_expression_return_value <doreturn LIBASSERT_COMMA \
             libassert_ret_lhs LIBASSERT_COMMA std::is_lvalue_reference<decltype(libassert_value)>::value> \
               (libassert_value, *std::launder(&libassert_decomposer)); \
-        ) LIBASSERT_IF(doreturn)(.value,) \
-        LIBASSERT_WARNING_PRAGMA_POP
+        LIBASSERT_WARNING_PRAGMA_POP \
+        ) LIBASSERT_IF(doreturn)(.value,)
 
 #ifdef NDEBUG
  #define LIBASSERT_ASSUME_ACTION LIBASSERT_UNREACHABLE;

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -1170,17 +1170,19 @@ using libassert::ASSERTION;
  // with decltype(auto) in an expression like decltype(auto) x = __extension__ ({...}).y;
  #define LIBASSERT_STMTEXPR(B, R) (__extension__ ({ B R }))
  #if LIBASSERT_IS_GCC
-  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
+  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_GCC \
      _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
      _Pragma("GCC diagnostic ignored \"-Wuseless-cast\"") // #49
+  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_CLANG
   #define LIBASSERT_WARNING_PRAGMA_PUSH_GCC _Pragma("GCC diagnostic push")
   #define LIBASSERT_WARNING_PRAGMA_POP_GCC _Pragma("GCC diagnostic pop")
   #define LIBASSERT_WARNING_PRAGMA_PUSH_CLANG
   #define LIBASSERT_WARNING_PRAGMA_POP_CLANG
  #else
-  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
+  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_CLANG \
      _Pragma("GCC diagnostic ignored \"-Wparentheses\"") \
      _Pragma("GCC diagnostic ignored \"-Woverloaded-shift-op-parentheses\"")
+  #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_GCC
   #define LIBASSERT_WARNING_PRAGMA_PUSH_GCC
   #define LIBASSERT_WARNING_PRAGMA_POP_GCC
   #define LIBASSERT_WARNING_PRAGMA_PUSH_CLANG _Pragma("GCC diagnostic push")
@@ -1193,7 +1195,8 @@ using libassert::ASSERTION;
  #define LIBASSERT_WARNING_PRAGMA_POP_CLANG
  #define LIBASSERT_WARNING_PRAGMA_PUSH_GCC
  #define LIBASSERT_WARNING_PRAGMA_POP_GCC
- #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA
+ #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_GCC
+ #define LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_CLANG
  #define LIBASSERT_STATIC_CAST_TO_BOOL(x) libassert::detail::static_cast_to_bool(x)
  namespace libassert::detail {
      template<typename T> constexpr bool static_cast_to_bool(T&& t) {
@@ -1267,11 +1270,13 @@ using libassert::ASSERTION;
 #endif
 #define ASSERT_INVOKE(expr, doreturn, check_expression, name, type, failaction, ...) \
         /* must push/pop out here due to nasty clang bug https://github.com/llvm/llvm-project/issues/63897 */ \
+        /* must do awful stuff to workaround differences in where gcc and clang allow these directives to go */ \
         LIBASSERT_WARNING_PRAGMA_PUSH_CLANG \
-        LIBASSERT_WARNING_PRAGMA_PUSH_GCC \
         LIBASSERT_IGNORE_UNUSED_VALUE \
-        LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA \
+        LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_CLANG \
         LIBASSERT_STMTEXPR( \
+          LIBASSERT_WARNING_PRAGMA_PUSH_GCC \
+          LIBASSERT_EXPRESSION_DECOMP_WARNING_PRAGMA_GCC \
           auto libassert_decomposer = \
                          libassert::detail::expression_decomposer(libassert::detail::expression_decomposer{} << expr); \
           LIBASSERT_WARNING_PRAGMA_POP_GCC \

--- a/tests/demo/demo.cpp
+++ b/tests/demo/demo.cpp
@@ -222,7 +222,7 @@ public:
         qux();
 
         {
-            M() < 2;
+            M() < 2; // NOLINT(clang-diagnostic-unused-comparison)
             puts("----");
             assert(M() < 2);
             puts("----");

